WIP: [Offload] Add testing for Offload program and kernel related entry points

callumfare opened the pull request.

callumfare committed:
WIP: Implement olMemAlloc, olMemFree

callumfare committed:
Add size check

callumfare committed:
Implement minimum Offload API needed to launch a SYCL kernel

callumfare committed:
Make a copy of the program binary in olCreateProgram

callumfare committed:
Rework kernel arguments

callumfare committed:
Update Offload unit tests

callumfare committed:
Kernel launch size arguments

callumfare committed:
Remove currently unused alignment param

callumfare committed:
Fix formatting

callumfare committed:
Fix leak in olReleaseQueue

callumfare committed:
General tidy up; improve documentation and formatting

callumfare committed:
Revert plugin changes

The offload unit tests will no longer work on host
Kernel execution will no longer work on CUDA

callumfare committed:
Rename `ol_*_handle_t_` -> `ol_*_impl_t`

callumfare committed:
Rework program and kernel implementation

callumfare committed:
WIP program and kernel launch testing

callumfare committed:
Various fixes to address review feedback

callumfare committed:
Formatting

callumfare committed:
Alternative memcpy implementation

jhuber6 reviewed (commented):
Some comments, mostly concerns about the proposed API.

callumfare commented on offload/liboffload/API/Enqueue.td:
```diff
@@ -0,0 +1,104 @@
+}
+
+def : Function {
+    let name = "olEnqueueMemcpyHtoD";
```
Yeah, since there seemed to be no objects to the new memcpy API you proposed I can remove these redundant functions

jhuber6 commented on offload/liboffload/API/Kernel.td:
```diff
@@ -0,0 +1,76 @@
+        "The created kernel has an initial reference count of 1."
+    ];
+    let params = [
+        Param<"ol_program_handle_t", "Program", "handle of the program", PARAM_IN>,
```
It might be nice to expose a read / write on the image before it's loaded, since that's what the image access lets us do. It's a very fast way to initialize globals.

I'll need to fix up some of the image handling in the plugins, right now it's really, really bad about managing lifetimes, just copying a pointer and assuming that the data lives forever (Even after global destructors have run).

callumfare commented on offload/liboffload/API/Kernel.td:
```diff
@@ -0,0 +1,76 @@
+}
+
+def : Function {
+    let name = "olSetKernelArgValue";
```
This is purely something that I've carried over from UR/SYCL/OpenCL. Somewhat out of habit but also because I was thinking of way down the line when we want to add an OpenCL plugin, having an API like this would make life easier.

That said, I do agree that this is a bit high-level and doesn't really match the low-level abstraction we're going for in the rest of the API.

One small advantage is that this lets the plugins/implementation lay out the argument struct according to whatever layout the underlying driver expects. On CUDA I believe arguments are just aligned to their size, is the same true on AMDGPU as well? Basically I'm wondering if we can explicitly say 'lay out the arguments this way' in the Offload API and know it will work.

callumfare replied:
Right, that makes sense. Having both the array-of-pointers way of passing arguments could be beneficial for the UR/SYCL use case, but passing the whole struct is easier to implement as things are so I'm happy to go that way for now.

callumfare commented on offload/liboffload/API/Memory.td:
```diff
@@ -0,0 +1,48 @@
+  let desc = "Frees a memory allocation previously made by olMemAlloc";
+  let params = [
+    Param<"ol_device_handle_t", "Device", "handle of the device to allocate on", PARAM_IN>,
+    Param<"ol_alloc_type_t", "Type", "type of the allocation", PARAM_IN>,
```
Fully agree with this, and actually I don't know if a hashmap is really necessary. On CUDA we can use `cuPointerGetAttributes` with `CU_POINTER_ATTRIBUTE_IS_MANAGED` and `CU_POINTER_ATTRIBUTE_MEMORY_TYPE` to know whether to use `cuMemFree` or `cuMemFreeHost`. If it's possible with hsa as well we could drop the kind from the plugin interface. I didn't tackle that in this PR since it seemed like a fairly invasive plugin change.

callumfare commented on offload/liboffload/API/Kernel.td:
```diff
@@ -0,0 +1,76 @@
+}
+
+def : Function {
+    let name = "olRetainKernel";
```
This is a consequence of the ability to set individual (or all) kernel arguments on a kernel object. With that, the kernel becomes the global address + whatever argument state the user has set, which necessitates tracking this state.

If we go the direction of passing arguments at the time of a kernel launch then we can drop all this.

callumfare committed:
Various fixes

callumfare commented:
@jhuber6 I've addressed some of your comments on here on the original PR (#122106). After fixing an issue with the __tgt_device_image lifetime I've realised I no longer need the plugin changes, so I'd like to roll these two PRs into one, since having reviews split across both is confusing. I'll close this PR soon and move the changes across.

jhuber6 commented:
> @jhuber6 I've addressed some of your comments on here on the original PR (#122106). After fixing an issue with the __tgt_device_image lifetime I've realised I no longer need the plugin changes, so I'd like to roll these two PRs into one, since having reviews split across both is confusing. I'll close this PR soon and move the changes across.

That really needs to be fixed regardless, registering an image with the runtime should do a copy of all the data.

callumfare commented on offload/liboffload/API/Enqueue.td:
```diff
@@ -0,0 +1,104 @@
+        "At least one device must be a non-host device"
+    ];
+    let params = [
+        Param<"ol_queue_handle_t", "Queue", "handle of the queue", PARAM_IN>,
```
The queue is the first param in every function except `olCreateQueue` where it's the last, because it's an output parameter. Generally I'd prefer to keep all output pointers as the final argument in every function, but that's just personal preference so could be changed.

callumfare commented on offload/liboffload/API/Enqueue.td:
```diff
@@ -0,0 +1,104 @@
+//===----------------------------------------------------------------------===//
+
+def : Function {
+    let name = "olEnqueueMemcpy";
```
Off the top of my head I can't think of an reason why we couldn't make the queue handles optional.

In UR we have optional handles, and don't hide the fact that they're pointers so they can be set to null.

But if we want to avoid that then I can make the change to `olMemcpy` and `olMemcpyAsync`

jhuber6 replied:
I prefer `olMemcpy` if we're not making the distinction. Forcing the user to create a queue is fine since this is supposed to be lower level.

callumfare committed:
Rework test device code

jhuber6 commented on offload/unittests/OffloadAPI/common/Environment.cpp:
```diff
@@ -94,3 +95,53 @@ ol_platform_handle_t TestEnvironment::getPlatform() {
+  std::ifstream SourceFile;
+  SourceFile.open(SourcePath, std::ios::binary | std::ios::in | std::ios::ate);
+
+  if (!SourceFile.is_open()) {
```
`llvm::MemoryBuffer::getFileOrSTDIN` is much easier to use and is backed by `mmap()` i believe.

callumfare committed:
Change olFinishQueue -> olWaitQueue

callumfare committed:
Remove redundant olEnqueueMemcpy functions
